In [21]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
# Load the dataset


In [22]:
from sklearn.datasets import fetch_california_housing

data = fetch_california_housing()

In [23]:
#Step 1: Create a DataFrame from the dataset
df = pd.DataFrame(data.data, columns=data.feature_names)
df['price'] = data.target

In [24]:
df.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,price
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


In [25]:
#Step 2: Check for duplicates
df.duplicated().sum()

0

When im checking the duplicates I got - " 0 " . The dataset is clcean , now rows need to be reoved for this reason.

In [26]:
#Step3 : Handle the 500k Cap
rows = df[df['price'] >= 4.99]
print(f"Total rows with Price >= $500K: {len(rows)}")

df_clean = df[df['price'] < 4.99]
print(f"before: {df.shape}")
print(f"after: {df_clean.shape}")


Total rows with Price >= $500K: 994
before: (20640, 9)
after: (19646, 9)


Problem : 
* The dataset as artifical cap - any house worth more than 500k is was recorded as exacly 500k
* so they are not real data 

why moed hurts :
* If dont remove fake /incomplete data my model will learn from that data . And it will always unprediact expensive houses 

My Decision :
* we remove them so model only learns from accurate and trusty dataa
* At the end we lose 965 rows but matching 19,000 rows are cleaner and perfect .

In [27]:
# Step 4: Remove AveRooms and AveOccup outliers

df_clean = df_clean[df_clean['AveRooms'] < 50]
df_clean = df_clean[df_clean['AveOccup'] < 20]
print(f"After {df_clean.shape}")

After (19629, 9)


After removing $500k capped rows and outliers in AveRooms and Aveoccup, dataset reduced from 20,640 to 19,629 rows. Removd 1,011 rows (about 5%)".

In [28]:
#step 5 : Pratice missing values

print(df_clean.isnull().sum())

MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
price         0
dtype: int64


* This is pratice only - there are no missing values in this dataset 
* if there were , we weould fill them like this 
df_clean = df_clean.fillna(df_clean.median())
